widgets/src/input.rs

```diff
@@ -116,366 +71,103 @@ impl TextArea {
     self.caret = caret;
   }
 }
 
-#[derive(Debug)]
-struct PreEditState {
-  position: usize,
-  value: Option<String>,
-}
-
-struct ImeHandle<H> {
-  host: H,
-  pre_edit: Option<PreEditState>,
-  guard: Option<SubscriptionGuard<BoxSubscription<'static>>>,
-  window: Sc<Window>,
-  caret_id: TrackId,
-}
-
-impl<E, H> ImeHandle<H>
-where
-  E: EditableText + 'static,
-  H: StateWriter<Value = E>,
-{
-  fn new(window: Sc<Window>, host: H, caret_id: TrackId) -> Self {
-    Self { window, host, pre_edit: None, guard: None, caret_id }
-  }
-  fn ime_allowed(&mut self) {
-    self.window.set_ime_allowed(true);
-    self.track_cursor();
-  }
-
-  fn ime_disallowed(&mut self) {
-    self.window.set_ime_allowed(false);
-    self.guard = None;
-  }
-
-  fn update_pre_edit(&mut self, e: &ImePreEditEvent) {
-    match &e.pre_edit {
-      ImePreEdit::Begin => {
-        let mut host = self.host.write();
-        let rg = host.caret().select_range();
-        host.writer().delete_byte_range(&rg);
-        self.pre_edit = Some(PreEditState { position: rg.start, value: None });
-      }
-      ImePreEdit::PreEdit { value, cursor } => {
-        let Some(PreEditState { position, value: edit_value }) = self.pre_edit.as_mut() else {
-          return;
-        };
-        let mut host = self.host.write();
-        let mut writer = host.writer();
-        if let Some(txt) = edit_value {
-          writer.delete_byte_range(&(*position..*position + txt.len()));
-        }
-        writer.insert_str(value);
-        writer.set_to(*position + cursor.map_or(0, |(start, _)| start));
-        *edit_value = Some(value.clone());
-      }
-      ImePreEdit::End => {
-        if let Some(PreEditState { value: Some(txt), position, .. }) = self.pre_edit.take() {
-          let mut host = self.host.write();
-          let mut writer = host.writer();
-          writer.delete_byte_range(&(position..position + txt.len()));
-        }
-      }
-    }
-    if self.pre_edit.is_none() {
-      self.track_cursor();
-    } else {
-      self.guard = None;
-    }
-  }
-
-  fn track_cursor(&mut self) {
-    if self.guard.is_some() {
-      return;
-    }
-
-    let window = self.window.clone();
-    let caret_id = self.caret_id.clone();
-    let tick_of_layout_ready = window
-      .frame_tick_stream()
-      .filter(|msg| matches!(msg, FrameMsg::LayoutReady(_)));
-    self.guard = Some(
-      self
-        .host
-        .modifies()
-        .sample(tick_of_layout_ready)
-        .box_it()
-        .subscribe(move |_| {
-          if let Some(wid) = caret_id.get() {
-            let pos = window.map_to_global(Point::zero(), wid);
-            let size = window.widget_size(wid).unwrap_or_default();
-            window.set_ime_cursor_area(&Rect::new(pos, size));
-          }
-        })
-        .unsubscribe_when_dropped(),
-    );
-  }
-}
-
 impl ComposeChild<'static> for Input {
-  type Child = Option<Placeholder>;
+  type Child = Option<Widget<'static>>;
```

Comment:
I believe there is no need to include a placeholder in the basic input, as higher-level widgets can handle its composition in their logic.